astropy/wcs/wcsapi/utils.py

```diff
@@ -114,3 +116,188 @@ def wcs_info_str(wcs):
+        if matches == 0:
+            raise ValueError("The world coordinate types of the two WCS do not match")
+        elif matches > 1:
+            unique_match = False
+            break
+        else:
+            mapping.append(classes2.index(class1))
+
+    if unique_match:
+
+        # Classes are unique, so we need to re-order matrix2 along the world
+        # axis using the mapping we found above.
+        matrix2 = matrix2[mapping]
+
+    elif classes1 != classes2:
+
+        raise ValueError("World coordinate order doesn't match and automatic matching is ambiguous")
+
+    matrix = np.matmul(matrix2.T, matrix1)
+
+    return matrix
+
+
+def split_matrix(matrix):
+    """
+    Given an axis correlation matrix from a WCS object, return information about
+    the individual WCS that can be split out.
+
+    The output is a list of tuples, where each tuple contains a list of
+    pixel dimensions and a list of world dimensions that can be extracted to
+    form a new WCS. For example, in the case of a spectral cube with the first
+    two world coordinates being the celestial coordinates and the third
+    coordinate being an uncorrelated spectral axis, the matrix would look like::
+
+        array([[ True,  True, False],
+               [ True,  True, False],
+               [False, False,  True]])
+
+    and this function will return ``[([0, 1], [0, 1]), ([2], [2])]``.
+    """
+
+    pixel_used = []
+
+    split_info = []
+
+    for ipix in range(matrix.shape[1]):
+        if ipix in pixel_used:
+            continue
+        pixel_include = np.zeros(matrix.shape[1], dtype=bool)
+        pixel_include[ipix] = True
+        n_pix_prev, n_pix = 0, 1
+        while n_pix > n_pix_prev:
+            world_include = matrix[:, pixel_include].any(axis=1)
+            pixel_include = matrix[world_include, :].any(axis=0)
+            n_pix_prev, n_pix = n_pix, np.sum(pixel_include)
+        pixel_indices = list(np.nonzero(pixel_include)[0])
+        world_indices = list(np.nonzero(world_include)[0])
+        pixel_used.extend(pixel_indices)
+        split_info.append((pixel_indices, world_indices))
+
+    return split_info
+
+
+def pixel_to_pixel(wcs_in, wcs_out, *inputs):
+    """
+    Transform pixel coordinates in a dataset with a WCS to pixel coordinates
+    in another dataset with a different WCS. This function is designed to
+    efficiently deal with input pixel arrays that are broadcasted views of
+    smaller arrays.
+
+    Parameters
+    ----------
+    wcs_in : `~astropy.wcs.wcsapi.BaseHighLevelWCS`
+        A WCS object for the original dataset which complies with the
+        high-level shared APE 14 WCS API.
+    wcs_out : `~astropy.wcs.wcsapi.BaseHighLevelWCS`
+        A WCS object for the target dataset which complies with the
+        high-level shared APE 14 WCS API.
+    *inputs :
+        Scalars or arrays giving the pixel coordinates to transform.
+    """
+
+    # Shortcut for scalars
+    if np.isscalar(inputs[0]):
```

Comment:
Is this supposed to work for WCS with 1 input and 1 output, e.g. spectral only? If so, perhaps add also a check for the dimensionality.

Reply:
This is only for scalars - for the 1D case one might still pass in broadcasted arrays that can be transformed more efficiently. This is covered by the new test I just added.